### PR TITLE
feat: implement noEmptyDocumentation

### LINKED ISSUE #7696: 📎 Port `comment-no-empty` from Stylelint
### Description

https://stylelint.io/user-guide/rules/comment-no-empty/

Should we also implement this for other languages?

Recommended naming `noEmptyComment`

### PR DESCRIPTION
**Pull request by Netail**
## Summary

Port stylelint's [comments-no-empty](https://stylelint.io/user-guide/rules/comment-no-empty/) and added support for other languages

Closes #7696

## Test Plan

Unit tests

### CODE REVIEW

File: .changeset/wise-hoops-dream.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+---
+"@biomejs/biome": patch
+---
+
+Added the new nursery rule [`noEmptyDocumentation`](https://biomejs.dev/linter/rules/no-empty-documentation), which disallows empty documentation (comments & descriptions) in HTML, CSS, JavaScript, JSON & GraphQL.
```
Comment (ematipico) [inline .changeset/wise-hoops-dream.md]:
```suggestion
Added the new nursery rule [`noEmptyDocumentation`](https://biomejs.dev/linter/rules/no-empty-documentation), which disallows empty documentation (comments & descriptions) in HTML, CSS, JavaScript, JSONC & GraphQL.
```

By default JSON doesn't support comments

File: crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        }))
+    }
+}
+
+fn leading_comments(syntax_node: &SyntaxNode<CssLanguage>) -> Vec<SyntaxTriviaPiece<CssLanguage>> {
+    if let Some(token) = syntax_node.first_token() {
+        token
+            .leading_trivia()
+            .pieces()
+            .filter(|piece| piece.is_comments())
```
Comment (ematipico) [inline crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: In JavaScript we have the concept of multiline comments. I suggest using that to filter trivia

File: crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            .collect()
+    } else {
+        Vec::new()
+    }
+}
+
+fn trailing_comments(syntax_node: &SyntaxNode<CssLanguage>) -> Vec<SyntaxTriviaPiece<CssLanguage>> {
```
Comment (ematipico) [inline crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: Generally these kinds of comments are leading comments, so I would exclude trailing comments.

If you found trialing comments that match the criteria, add them in the tests

File: crates/biome_graphql_analyze/src/lint/nursery/no_empty_documentation.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+use biome_console::markup;
+use biome_graphql_syntax::{GraphqlDescription, GraphqlLanguage, GraphqlRoot};
+use biome_rowan::{AstNode, SyntaxNode, SyntaxTriviaPiece, TextRange};
+use biome_rule_options::no_empty_documentation::NoEmptyDocumentationOptions;
+
+declare_lint_rule! {
+    /// Disallow empty documentation.
+    ///
+    /// Enforces that comments are not empty. This helps maintain code quality by preventing meaningless
+    /// or placeholder comments that don't provide any documentation value.
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_empty_documentation.rs]: Empty comments and empty documentation are different though, semantically speaking. The rule swaps their meaning, which isn't correct 

```js
let foo; // this is a comment

/** This is docs **/
let bar;
```

**Review by ematipico — CHANGES_REQUESTED** [review]
The business logic of the rule can be shared in one single type, maybe in the analyser, or rule options crates.

As for the solution, create a struct that implements `Iterator` and return `SyntaxTokenKind` as `Item`, so that callers can use the `filter` function, which you're using in each different crate.

I would also revisit docs and business logic. I believe trailing comments shouldn't be part of the logic, however if we do think we should, we should add tests because at the moment there are none.

File: crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs (hunk 2 above)
Comment (Netail) [reply crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]:
```
// Hello
//
// Something else, but part of the full block
```

Something like this you mean?

Comment (ematipico) [reply crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: No.

// -> single line comment 
/***/ -> multi line comment 

Then again, it's all part of the business logic of the function and the misalignment between comment and documentation

Comment (Netail) [reply crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: Ahh okay, but that one I got covered, both `/* */` & `/** */`

File: crates/biome_analyze/src/trivia.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+/// Iterates leading trivia comments for a root and all its descendants.
+pub struct LeadingCommentTriviaPieces<L: Language> {
+    nodes: std::vec::IntoIter<SyntaxNode<L>>,
+    current_pieces: Option<SyntaxTriviaPiecesIterator<L>>,
+}
+
+impl<L: Language> LeadingCommentTriviaPieces<L> {
+    pub fn new(root: &SyntaxNode<L>) -> Self {
+        let mut nodes = Vec::new();
```
Comment (ematipico) [inline crates/biome_analyze/src/trivia.rs]: Nit: you could add a `debug_assert`, and check for `root.parent()`. If it exists, it's not a root node

File: crates/biome_analyze/src/trivia.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        nodes.push(root.clone());
+        nodes.extend(root.descendants());
```
Comment (ematipico) [inline crates/biome_analyze/src/trivia.rs]: Descendants already include the current node (that's why you see `skip(1)` all the time)

File: crates/biome_analyze/src/trivia.rs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                    return Some(piece);
+                }
+
+                self.current_pieces = None;
+            }
+
+            let node = self.nodes.next()?;
+            self.current_pieces = node
+                .first_token()
+                .map(|token| token.leading_trivia().pieces());
```
Comment (ematipico) [inline crates/biome_analyze/src/trivia.rs]: Nit: Early return if the first token doesn't exist

File: crates/biome_analyze/src/trivia.rs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+impl<L: Language> Iterator for LeadingCommentTriviaPieces<L> {
+    type Item = SyntaxTriviaPiece<L>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if let Some(pieces) = &mut self.current_pieces {
+                if let Some(piece) = pieces.find(|piece| piece.is_comments()) {
+                    return Some(piece);
+                }
+
```
Comment (ematipico) [inline crates/biome_analyze/src/trivia.rs]: This is incorrect. You must exhaust the iterator. If there are 5 comments here, you'll return only the first one, the other four are lost.

Your tests assume only and always one leading comment. What about 

```css
/***/
/***/
.css {}
```

File: crates/biome_analyze/src/trivia.rs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+                self.current_pieces = None;
```
Comment (ematipico) [inline crates/biome_analyze/src/trivia.rs]: This should be set to `None` only after the iterator is exhausted

File: crates/biome_analyze/src/trivia.rs [hunk 10]
```diff
@@ -0,0 +1,41 @@
+use biome_rowan::{Language, SyntaxNode, SyntaxTriviaPiece, syntax::SyntaxTriviaPiecesIterator};
+
+/// Iterates leading trivia comments for a root and all its descendants.
+pub struct LeadingCommentTriviaPieces<L: Language> {
```
Comment (ematipico) [inline crates/biome_analyze/src/trivia.rs]: Nit: by convention, types that are meant to be used as iterator should the word `Iterator` in their name

File: crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let syntax_node = node.syntax();
+
+        let mut found: Vec<_> = LeadingCommentTriviaPieces::new(syntax_node)
+            .filter(|comment| is_empty_comment(comment.text().trim()))
+            .map(|comment| comment.text_range())
+            .collect();
+
+        found.sort_unstable_by_key(|range| (range.start(), range.end()));
```
Comment (ematipico) [inline crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: Nit: Don't sort if it's empty

File: crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        found.dedup();
```
Comment (ematipico) [inline crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: What's this for?

File: crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        language: "css",
+        recommended: false,
+        sources: &[RuleSource::Stylelint("comment-no-empty").same()],
+    }
+}
+
+impl Rule for NoEmptyDocumentation {
+    type Query = Ast<AnyCssRoot>;
+    type State = Vec<TextRange>;
+    type Signals = Option<Self::State>;
```
Comment (ematipico) [inline crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: This is wrong. The signal should either return Option or Vec, not a mix of both.

Considering the rule, you want to return `Vec<State>` where `State` is `TextRange`

This is covered in the contribution guide

**Review by ematipico — CHANGES_REQUESTED** [review]
I found some bugs in the implementation. We should add more tests.

Also, we must add tests for JSX too. They have comments too

File: crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs (hunk 13 above)
Comment (Netail) [reply crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: I don't think that's allowed by the Rule type;

```
type mismatch resolving `<Vec<TextRange> as IntoIterator>::Item == Vec<TextRange>`
expected struct `std::vec::Vec<biome_rowan::TextRange>`
   found struct `biome_rowan::TextRange`
```

Looking at other rules with a `Vec<>` they also seem to return in an Option

File: crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs (hunk 12 above)
Comment (Netail) [reply crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: For some reason, the walking seems to catch certain trivia multiple times. So I try to filter out duplicates

Comment (ematipico) [reply crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: That shouldn't happen, maybe there's a bug in the iterator? It would be useful to add some unit tests

File: crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs (hunk 13 above)
Comment (ematipico) [reply crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: State become TextRange, Signals becomes Vec<State>

Comment (ematipico) [reply crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: Concrete examples 

https://github.com/biomejs/biome/blob/main/crates%2Fbiome_js_analyze%2Fsrc%2Flint%2Fsuspicious%2Fno_ts_ignore.rs#L53-L54

Comment (Netail) [reply crates/biome_css_analyze/src/lint/nursery/no_empty_documentation.rs]: Ohhhh like that, thanks!

**Comment by Netail** [comment]
> Also, we must add tests for JSX too. They have comments too

For some reason, those only seem to be caught with trailing trivia, so I added collecting the trailing trivia as well again

**Comment by ematipico** [comment]
> > Also, we must add tests for JSX too. They have comments too
> 
> For some reason, those only seem to be caught with trailing trivia, so I added collecting the trailing trivia as well again

Yet you still have the option; just because JSX needs the trailing trivia, doesn't mean that all the other languages have to do the same.

The iterator can accept an option at creating to include the trailing trivia

**Comment by Netail** [comment]
> > > Also, we must add tests for JSX too. They have comments too
> > 
> > 
> > For some reason, those only seem to be caught with trailing trivia, so I added collecting the trailing trivia as well again
> 
> Yet you still have the option; just because JSX needs the trailing trivia, doesn't mean that all the other languages have to do the same.
> 
> The iterator can accept an option at creating to include the trailing trivia

Which languages/dialects do you recommend to only use trailing? JSX, JSONC, Graphql?

**Comment by ematipico** [comment]
I don't know the CST by heart, but based on your tests and the intention of the rule, I believe we should always target leading trivia, and JSX comments should be excluded for the time being. I've never seen documentation comments between JSX tags

**Comment by Netail** [comment]
> I've never seen documentation comments between JSX tags

Really? It's pretty common internally for us. Usually with the reasoning on why a certain comment has been commented out.
And for GraphQL is pretty common to add `# Some documentation` behind a field in a type definition

**Comment by ematipico** [comment]
> Really? It's pretty common internally for us. Usually with the reasoning on why a certain comment has been commented out.

I still think we have two separate definitions of what "documentation" means, and that's the reason why the review is becoming longer than it should be. 

The original rule is about empty **comments**, but the rule implemented in Biome is called **documentation**. To me, these are two separate things:
- Documentations are comments
- Not all comments are documentation

When you explain "why you did something in the code", that's just a comment. When you put a comment before a function, that's documentation. etc.

**Comment by Netail** [comment]
> > Really? It's pretty common internally for us. Usually with the reasoning on why a certain comment has been commented out.
> 
> I still think we have two separate definitions of what "documentation" means, and that's the reason why the review is becoming longer than it should be.
> 
> The original rule is about empty **comments**, but the rule implemented in Biome is called **documentation**. To me, these are two separate things:
> 
> * Documentations are comments
> * Not all comments are documentation
> 
> When you explain "why you did something in the code", that's just a comment. When you put a comment before a function, that's documentation. etc.

Maybe the naming of the rule is a bit confusing and it's scope. The original Stylelint rule was about all comments, so it was also nice to support the rule in other languages. But because GraphQL also has something called [descriptions](https://graphql.org/learn/schema/#descriptions), which are kind of similar, but not comments. (They are regular nodes instead of trivia) And to also cover those, it didn't really make sense to name the rule `noEmptyComments`. So the "documentation" refers to an overarching of comments & graphql descriptions, instead of referring to the subset of comment types

**Comment by ematipico** [comment]
IMHO we should have different rules. For example, JS has two styles of comments too, and the multiline **can** be JsDocs. JSON has comments only if configured. 

Different varieties, it's very hard to find the line.
